kratos/modified_shape_functions/triangle_2d_3_ausas_incised_shape_functions.cpp

```diff
@@ -0,0 +1,152 @@
+    Triangle2D3AusasModifiedShapeFunctions(pInputGeometry, rNodalDistancesWithExtrapolated), mExtraEdgeRatios(rExtrapolatedEdgeRatios) {};
+
+/// Destructor
+Triangle2D3AusasIncisedShapeFunctions::~Triangle2D3AusasIncisedShapeFunctions() {};
+
+/// Turn back information as a string.
+std::string Triangle2D3AusasIncisedShapeFunctions::Info() const {
+    return "Triangle2D3N Ausas incised shape functions computation class.";
+}
+
+/// Print information about this object.
+void Triangle2D3AusasIncisedShapeFunctions::PrintInfo(std::ostream& rOStream) const {
+    rOStream << "Triangle2D3N Ausas incised shape functions computation class.";
+}
+
+/// Print object's data.
+void Triangle2D3AusasIncisedShapeFunctions::PrintData(std::ostream& rOStream) const {
+    const GeometryPointerType p_geometry = this->GetInputGeometry();
+    const Vector nodal_distances = this->GetNodalDistances();
+    rOStream << "Triangle2D3N Ausas incised shape functions computation class:\n";
+    rOStream << "\tGeometry type: " << (*p_geometry).Info() << "\n";
+    std::stringstream distances_buffer;
+    std::ostringstream stm;
+    for (unsigned int i = 0; i < nodal_distances.size(); ++i) {
+        stm << nodal_distances(i);
+        distances_buffer << stm.str() << " ";
+    }
+    rOStream << "\tExtrapolated distance values: " << distances_buffer.str();
+}
+
+// Returns the nodal distances vector.
+const Vector& Triangle2D3AusasIncisedShapeFunctions::GetExtrapolatedEdgeRatios() const {
+    return mExtraEdgeRatios;
+}
+
+// Sets the condensation matrix to transform the subdivsion positive side values to entire element ones.
+void Triangle2D3AusasIncisedShapeFunctions::SetPositiveSideCondensationMatrix(
+    Matrix& rPosSideCondMatrix,
+    const std::vector<int>& rEdgeNodeI,
+    const std::vector<int>& rEdgeNodeJ,
+    const std::vector<int>& rSplitEdges)
+{
+    const std::size_t n_nodes = 3;
+    const std::size_t n_edges = 3;
+
+    // Initialize intersection points condensation matrix
+    rPosSideCondMatrix = ZeroMatrix(n_nodes + n_edges, n_nodes);
+
+    // Get the nodal distances vector
+    const Vector& nodal_distances = this->GetNodalDistances();
+
+    // Fill the original geometry points main diagonal
+    for (std::size_t i = 0; i < n_nodes; ++i) {
+        rPosSideCondMatrix(i,i) = (nodal_distances(i) > 0.0) ? 1.0 : 0.0;
+    }
+
+    // Compute the intersection points contributions
+    for (std::size_t id_edge = 0; id_edge < n_edges; ++id_edge) {
+        // Check if the edge has an intersection point
+        if (rSplitEdges[n_nodes+id_edge] != -1) {
+            // Get the nodes that compose the edge
+            const std::size_t edge_node_i = rEdgeNodeI[id_edge];
+            const std::size_t edge_node_j = rEdgeNodeJ[id_edge];
+
+            // Transform definition of edge ID from divide_triangle_2d_3.h to definition of triangle_2d_3.h (geometry)
+            unsigned int ratio_edge_id = (id_edge+2)%3;
```

Comment:
This relation is a coincidence. I decided to use it, so the code is faster, although I am not sure it really is, and maybe it would be better and clearer to have transformation member arrays, like for the tetrahedra class. What do you think?

Reply:
I'd have the transformation arrays as we have for the tetrahedra class for the sake of clarity .